src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -193,48 +202,84 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
      * </ol>
      *
      * @param programmingExercise The programmingExercise that should be setup
-     * @param isImportedFromFile  defines if the programming exercise is imported from a file
+     * @param isImportedFromFile  defines if the programming exercise is imported from a file, if the
+     *                                exercise is imported, the build plans will not be triggered to prevent erroneous builds
      * @return The new setup exercise
      * @throws GitAPIException If something during the communication with the remote Git repository went wrong
      * @throws IOException     If the template files couldn't be read
      */
-    @Transactional // TODO: apply the transaction on a smaller scope
-    // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
     public ProgrammingExercise createProgrammingExercise(ProgrammingExercise programmingExercise, boolean isImportedFromFile) throws GitAPIException, IOException {
-        programmingExercise.generateAndSetProjectKey();
-        final User exerciseCreator = userRepository.getUser();
+        AtomicBoolean gitExceptionThrown = new AtomicBoolean(false);
 
+        final User exerciseCreator = userRepository.getUser();
         VersionControlService versionControl = versionControlService.orElseThrow();
+
+        programmingExercise.generateAndSetProjectKey();
         programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
-        programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
-        initParticipations(programmingExercise);
-        setURLsAndBuildPlanIDsForNewExercise(programmingExercise);
 
-        // Save participations to get the ids required for the webhooks
-        connectBaseParticipationsToExerciseAndSave(programmingExercise);
+        ProgrammingExercise createdExercise = transactionTemplate.execute(new TransactionCallback<ProgrammingExercise>() {
+
+            @Override
+            public ProgrammingExercise doInTransaction(@NotNull TransactionStatus status) {
```

Comment:
Please replace this with a lambda or even better with a method reference for better readability.

Reply:
If I put the lines within the `doInTransaction` method in another method and just called it in `doInTransaction`, I could not access the attributes set outside the transactional scope. So I would have to use many method attributes. I don't know what to prefer.